api/src/pcapi/utils/cache.py

```diff
@@ -0,0 +1,146 @@
+from functools import partial
+from functools import wraps
+from hashlib import sha256
+import json
+from typing import Any
+from typing import Callable
+from typing import Iterable
+from typing import cast
+
+from flask import current_app
+import pydantic
+
+
+class _CacheProxy:
+    def __init__(self, data: str):
+        self._json = data
+        self._data = None
+
+    def __getattr__(self, name: str) -> Any:
+        if self._data is None:
+            self._data = json.loads(self._json)
+        return self._data.get(name)  # type: ignore [attr-defined]
+
+    def json(self, *args: Iterable[Any], **kwargs: dict[str, Any]) -> str:
+        return self._json
+
+
+def get_from_cache(
+    retriever: Callable[..., pydantic.BaseModel | str],
+    key_template: str,
+    key_args: tuple[str, ...] | dict[str, Any] | None = None,
+    expire: int | None = 60 * 60 * 24,  # 24h
+    return_type: type = str,
+    force_update: bool = False,
+) -> pydantic.BaseModel | str:
+    """
+    Retrieve data from cache if available else use the retriever callable to retrieve data and store it in cache.
+
+    :param key_template: A template to generate the key. This must be compatible with printf style string formatting
+        see for documentation: https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting
+    :param key_args: `list` or `dict` of `str` to fill the template.
+    :param retriever: Function to call to retrive the data if they are not in the cache. As it does not take args you
+        can use functools.partial to prefil its arguments. It must return a pydantic.BaseModel child or str.
+    :param expire: The number of seconds before the cache expire. If None it will be set to never expire.
+        It's default value is 86400 (24h).
+    :param return_type: Type awaited for return value. This is meant to fool mypy and spectree_serialize and keep
+        compatibility. It can be either `BaseModel` or `str`.
+    :param force_update: If True force the update of the field cache.
+    """
+    redis_client = current_app.redis_client  # type: ignore [attr-defined]
+    miss = False
+    if key_args:
+        key = key_template % key_args
+    else:
+        key = key_template
+
+    data = redis_client.get(key)
+    if not data and not redis_client.exists(key):
+        miss = True
+
+    if miss or force_update:
+        data = retriever()
+        if isinstance(data, pydantic.BaseModel):
+            data = data.json(exclude_none=False, by_alias=True)
+        if expire is not None:
+            redis_client.set(key, data.encode("utf-8"), ex=expire)
+        else:
+            redis_client.set(key, data.encode("utf-8"))
```

Comment:
(datail) Ne peut-on pas simplifier en : 

```suggestion
        redis_client.set(key, data.encode("utf-8"), ex=expire)
```